range-v3 build errors with MSVC 14.25

range-v3 master fails to build with latest MSVC 14.25 ( released a few days ago, VS 16.5 ) due to multiple compiler errors

the issue can be reproduced with the following code
```
#include <range/v3/view/iota.hpp>
#include <range/v3/view/take.hpp>
#include <range/v3/view/reverse.hpp>

int main()
{
    using namespace ranges;
    auto rng5 = views::iota(10) | views::take(10) | views::reverse;
}
```

```
1>D:\Projects\range-v3\include\range/v3/utility/optional.hpp(579,1): error C2280: 'bool concepts::detail::operator &&(concepts::detail::boolean,bool)': attempting to reference a deleted function
1>D:\Projects\range-v3\include\concepts/concepts.hpp(751): message : see declaration of 'concepts::detail::operator &&'
1>D:\Projects\range-v3\include\concepts/concepts.hpp(751,25): message : 'bool concepts::detail::operator &&(concepts::detail::boolean,bool)': function was explicitly deleted
1>D:\Projects\range-v3\include\range/v3/utility/optional.hpp(909): message : see reference to class template instantiation 'ranges::optional<T>' being compiled
1>        with
1>        [
1>            T=ranges::counted_iterator<ranges::basic_iterator<ranges::iota_view<int,ranges::unreachable_sentinel_t>::cursor>,std::true_type,0>
1>        ]
1>D:\Projects\range-v3\include\range/v3/view/reverse.hpp(48): message : see reference to class template instantiation 'ranges::detail::non_propagating_cache<ranges::counted_iterator<ranges::basic_iterator<ranges::iota_view<int,ranges::unreachable_sentinel_t>::cursor>,std::true_type,0>,ranges::reverse_view<ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t>>>,true>' being compiled
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.25.28610\include\type_traits(1610): message : see reference to class template instantiation 'ranges::reverse_view<ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t>>>' being compiled
1>C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.25.28610\include\type_traits(1610): message : while compiling class template member function 'ranges::reverse_view<ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t>>> std::_Invoker_functor::_Call<_Ty,ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t>>>(_Callable &&,ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t>> &&) noexcept(<expr>)'
1>        with
1>        [
1>            _Ty=ranges::views::reverse_fn,
1>            _Callable=ranges::views::reverse_fn
1>        ]
1>D:\Projects\range-v3\include\range/v3/functional/concepts.hpp(33): message : see reference to variable template 'const bool is_invocable_v<ranges::views::reverse_fn,ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t> > >' being compiled
1>D:\Projects\range-v3\include\range/v3/view/view.hpp(76): message : see reference to variable template 'const bool invocable<ranges::views::reverse_fn,ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t> > >' being compiled
1>D:\Projects\range-v3\include\range/v3/view/view.hpp(83): message : see reference to variable template 'const bool invocable_view_closure<ranges::views::reverse_fn,ranges::take_view<ranges::iota_view<int,ranges::unreachable_sentinel_t> > >' being compiled
1>D:\Projects\range-v3\include\concepts/concepts.hpp(869): message : see reference to function template instantiation 'auto ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>::operator ()<std::true_type>(std::true_type) const' being compiled
1>D:\Projects\range-v3\test\view\test.cpp(8): message : see reference to function template instantiation 'concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>::operator bool(void) noexcept const<true>' being compiled
1>D:\Projects\range-v3\include\concepts/concepts.hpp(792): message : see reference to function template instantiation 'concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>::operator bool(void) noexcept const<true>' being compiled
1>D:\Projects\range-v3\include\concepts/concepts.hpp(796,34): message : while evaluating constexpr function 'concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>::impl'
1>D:\Projects\range-v3\include\concepts/concepts.hpp(796,41): message : while evaluating constexpr function 'concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>::operator bool'
1>D:\Projects\range-v3\include\concepts/concepts.hpp(795): message : while compiling class template member function 'concepts::detail::and_<concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>,std::true_type>::operator bool(void) noexcept const'
1>D:\Projects\range-v3\test\view\test.cpp(8): message : see reference to class template instantiation 'concepts::detail::and_<concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>,std::true_type>' being compiled
1>D:\Projects\range-v3\include\range/v3/utility/optional.hpp(579,1): error C2088: '&&': illegal for struct
1>D:\Projects\range-v3\include\concepts/concepts.hpp(796,34): message : while evaluating constexpr function 'concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>::impl'
1>D:\Projects\range-v3\include\concepts/concepts.hpp(796,41): message : while evaluating constexpr function 'concepts::detail::and_<concepts::detail::boolean_<ranges::defer::<lambda_2ae2609f2938c514d772d4c416ac2f09>>,concepts::detail::boolean_<ranges::defer::<lambda_4b7ebeada7fee5bba52ad0d317e38301>>>::operator bool'
```

--- a/include/concepts/concepts.hpp
+++ b/include/concepts/concepts.hpp
@@ -887,7 +887,7 @@ namespace concepts
 
         CPP_INLINE_VAR constexpr CPP_true_t CPP_true_{};
 
-        constexpr auto CPP_true(xNil) noexcept
+        constexpr auto CPP_true(xNil)
         {
             return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
         }
@@ -895,7 +895,7 @@ namespace concepts
         using xNil = Nil;
 #endif
 
-        constexpr auto CPP_true(Nil) noexcept
+        constexpr auto CPP_true(Nil)
         {
             return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
         }

--- a/include/concepts/concepts.hpp
+++ b/include/concepts/concepts.hpp
@@ -432,7 +432,7 @@
     /**/
 #else // ^^^ workaround / no workaround vvv
 #define CPP_broken_friend_member                                                \
-    template<std::true_type (&CPP_true)(::concepts::detail::xNil) =             \
+    template<::concepts::detail::boolean_true_t (&CPP_true)(::concepts::detail::xNil) = \
         ::concepts::detail::CPP_true>                                           \
     /**/
 #endif // CPP_WORKAROUND_MSVC_779763
@@ -870,16 +870,26 @@ namespace concepts
         struct Nil
         {};
 
+        struct _true_fn
+        {
+            std::true_type operator()(int) const noexcept
+            {
+                return {};
+            }
+        };
+
+        using boolean_true_t = boolean_<_true_fn>;
+
 #ifdef CPP_WORKAROUND_MSVC_779763
         enum class xNil {};
 
         struct CPP_true_t
         {
-            constexpr std::true_type operator()(Nil) const noexcept
+            constexpr boolean_true_t operator()(Nil) const noexcept
             {
                 return {};
             }
-            constexpr std::true_type operator()(xNil) const noexcept
+            constexpr boolean_true_t operator()(xNil) const noexcept
             {
                 return {};
             }
@@ -887,17 +897,17 @@ namespace concepts
 
         CPP_INLINE_VAR constexpr CPP_true_t CPP_true_{};
 
-        constexpr auto CPP_true(xNil)
+        constexpr boolean_true_t CPP_true(xNil)
         {
-            return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
+            return {};
         }
 #else
         using xNil = Nil;
 #endif
 
-        constexpr auto CPP_true(Nil)
+        constexpr boolean_true_t CPP_true(Nil)
         {
-            return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
+            return {};
         }
 
         template<typename T>

--- a/include/concepts/concepts.hpp
+++ b/include/concepts/concepts.hpp
@@ -746,20 +746,14 @@ namespace concepts
             return true;
         }
 
-        struct boolean
-        {
-            friend bool operator&&(boolean, bool) = delete;
-            friend bool operator&&(bool, boolean) = delete;
-            friend bool operator||(boolean, bool) = delete;
-            friend bool operator||(bool, boolean) = delete;
-        };
         template<typename T, typename U>
         struct and_;
         template<typename T, typename U>
         struct or_;
         template<typename T>
-        struct not_ : boolean
+        struct not_
         {
+            using _cpp_boolean_t = not_;
             constexpr operator bool() const noexcept
             {
                 return !(bool) T{};
@@ -769,20 +763,21 @@ namespace concepts
                 return T{};
             }
             template<typename That>
-            constexpr and_<not_, That> operator&&(That) const noexcept
+            constexpr and_<not_, typename That::_cpp_boolean_t> operator&&(That) const noexcept
             {
                 return {};
             }
             template<typename That>
-            constexpr or_<not_, That> operator||(That) const noexcept
+            constexpr or_<not_, typename That::_cpp_boolean_t> operator||(That) const noexcept
             {
                 return {};
             }
         };
 
         template<typename T, typename U>
-        struct and_ : boolean
+        struct and_
         {
+            using _cpp_boolean_t = and_;
             static constexpr bool impl(std::false_type) noexcept
             {
                 return false;
@@ -800,28 +795,21 @@ namespace concepts
                 return not_<and_>{};
             }
             template<typename That>
-            constexpr and_<and_, That> operator&&(That) const noexcept
+            constexpr and_<and_, typename That::_cpp_boolean_t> operator&&(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a conjunction should be "
-                    "defer:: concepts");
                 return {};
             }
             template<typename That>
-            constexpr or_<and_, That> operator||(That) const noexcept
+            constexpr or_<and_, typename That::_cpp_boolean_t> operator||(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a disjunction should be "
-                    "defer:: concepts");
                 return {};
             }
         };
 
         template<typename T, typename U>
-        struct or_ : boolean
+        struct or_
         {
+            using _cpp_boolean_t = or_;
             static constexpr bool impl(std::true_type) noexcept
             {
                 return true;
@@ -839,28 +827,21 @@ namespace concepts
                 return {};
             }
             template<typename That>
-            constexpr and_<or_, That> operator&&(That) const noexcept
+            constexpr and_<or_, typename That::_cpp_boolean_t> operator&&(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a disjunction should be "
-                    "defer:: concepts");
                 return {};
             }
             template<typename That>
-            constexpr or_<or_, That> operator||(That) const noexcept
+            constexpr or_<or_, typename That::_cpp_boolean_t> operator||(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a disjunction should be "
-                    "defer:: concepts");
                 return {};
             }
         };
 
         template<class Fn>
-        struct boolean_ : boolean
+        struct boolean_
         {
+            using _cpp_boolean_t = boolean_;
             boolean_() = default;
             constexpr boolean_(decltype(nullptr)) noexcept {}
             template<bool = true>
@@ -873,21 +854,13 @@ namespace concepts
                 return {};
             }
             template<typename That>
-            constexpr and_<boolean_, That> operator&&(That) const noexcept
+            constexpr and_<boolean_, typename That::_cpp_boolean_t> operator&&(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a disjunction should be "
-                    "defer:: concepts");
                 return {};
             }
             template<typename That>
-            constexpr or_<boolean_, That> operator||(That) const noexcept
+            constexpr or_<boolean_, typename That::_cpp_boolean_t> operator||(That) const noexcept
             {
-                static_assert(
-                    !META_IS_SAME(That, bool),
-                    "All expressions in a disjunction should be "
-                    "defer:: concepts");
                 return {};
             }
         };
@@ -914,17 +887,17 @@ namespace concepts
 
         CPP_INLINE_VAR constexpr CPP_true_t CPP_true_{};
 
-        constexpr std::true_type CPP_true(xNil)
+        constexpr auto CPP_true(xNil) noexcept
         {
-            return {};
+            return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
         }
 #else
         using xNil = Nil;
 #endif
 
-        constexpr std::true_type CPP_true(Nil)
+        constexpr auto CPP_true(Nil) noexcept
         {
-            return {};
+            return true ? nullptr : detail::make_boolean([](auto){return std::true_type{};});
         }
 
         template<typename T>

